lib/core/src/sockComm.cpp

```diff
@@ -280,6 +302,59 @@ irods::error readMsgBody(
 
 } // readMsgBody
 
+// NOLINTNEXTLINE(modernize-use-trailing-return-type)
+int set_socket_tcp_keepalive_options(int _sfd)
+{
+    int ec = 0;
+    rodsEnv env{};
+    if (ec = getRodsEnv(&env); ec < 0) {
+        return ec;
+    }
+
+    // Make sure that SO_KEEPALIVE option is set on the socket. If not, we need to enable it here because iRODS will use
+    // TCP keepalive for all socket connections over the network.
+    int optval = 0;
+    socklen_t optlen = sizeof(optval);
+    if (ec = getsockopt(_sfd, SOL_SOCKET, SO_KEEPALIVE, &optval, &optlen); ec < 0) {
+        return SOCKET_ERROR - errno;
+    }
+
+    if (0 == optval) {
+        optval = 1;
+        optlen = sizeof(optval);
+        if (ec = setsockopt(_sfd, SOL_SOCKET, SO_KEEPALIVE, &optval, optlen); ec < 0) {
+            return SOCKET_ERROR - errno;
+        }
+    }
+
+    // If the configuration is invalid for the TCP keepalive options, that option should not be set on the socket. This
+    // will allow the socket to use to the kernel configuration instead.
+    optval = env.tcp_keepalive_probes;
+    if (optval >= 0) {
+        optlen = sizeof(optval);
+        if (ec = setsockopt(_sfd, SOL_TCP, TCP_KEEPCNT, &optval, optlen); ec < 0) {
+            return SOCKET_ERROR - errno;
+        }
+    }
+
+    optval = env.tcp_keepalive_time;
+    if (optval >= 0) {
+        optlen = sizeof(optval);
+        if (ec = setsockopt(_sfd, SOL_TCP, TCP_KEEPIDLE, &optval, optlen); ec < 0) {
+            return SOCKET_ERROR - errno;
+        }
+    }
+
+    optval = env.tcp_keepalive_intvl;
+    if (optval >= 0) {
+        optlen = sizeof(optval);
+        if (ec = setsockopt(_sfd, SOL_TCP, TCP_KEEPINTVL, &optval, optlen); ec < 0) {
+            return SOCKET_ERROR - errno;
+        }
+    }
```

Comment:
I cannot find any information about the range of acceptable values for these, so if anybody knows, that would be nice.

The main thing that I have discovered is that 0 is not valid for `TCP_KEEPIDLE`. I suspect that this may be the case for `TCP_KEEPINTVL` and `TCP_KEEPCNT`, but I haven't tried it yet.

I went back and forth on whether non-negative integers was the right domain, but this seems like a strong case for limiting to positive integers.

Reply:
there's a spec / RFC somewhere...

Reply:
Alternatively, we could just try to apply whatever setting the user has provided and let it fail with an error. This is, essentially, exposing the TCP keepalive options for iRODS connections as configurable options, so we could just fall back on the behavior of `setsockopt` instead of worrying about it. Thoughts?

Reply:
Yes. As long as we're confident we're holding the OS API calls correctly, I think it's fine. No need to test OS APIs.

Reply:
I think so. This implementation is pulled almost directly from the example here: https://tldp.org/HOWTO/TCP-Keepalive-HOWTO/programming.html

I will leave a lengthy comment here in the code explaining why we don't check the values and some words in the documentation for these options warning the user to be careful.

Reply:
Sounds good.

Just so I understand, the logic you have is not going to change, correct?
You're going to add comments to help the next developer only.

Reply:
I see. And if they set it to a bad value `setsockopt` will report an error and the client will see that in the iRODS error code.

Carry on.

Reply:
Exactly

Reply:
Oops... It turns out that we can't do this approach because we need to set the members of the `rodsEnv` to *something* to indicate that they are not set, and I was using -1 for this purpose. I have confirmed that 0 is an invalid input for all 3 of these socket options, so we should only set the socket option when the value is a positive integer.

Also, I think that the default values in the documentation should be -1 so that "unset" is the default, but we will note what the defaults from the TCP spec are.

Thoughts?

Reply:
seems fair all around.

Reply:
No objections here.

Reply:
Done. Added words to docs to this effect, too